### PR TITLE
Add example with Cannon's Matrix Multiplication

### PR DESCRIPTION
**Pull request by ZwFink**
- Adds an implementation of Cannon's Matrix Multiplication algorithm to ```examples```
- Showcases the following features:
    - Use of Numba to accelerate computation
    - 2-dimensional chare arrays
    - Channels
    - Reductions
    - Futures

### CODE REVIEW

File: examples/cannon/cannon.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+    # create a dummy numba.njit decorator
+    def njit(func):
+        return func
+
+@njit
+def matmul(C, A, B):
+    C += A @ B
+
+class SubMatrix(Chare):
+    def __init__(self, subdim_size, charedim, init_done):
```
Comment (minitu) [inline examples/cannon/cannon.py:17]: I think this is missing a call to `super(SubMatrix, self).__init__(...)`, hence the LGTM warning.

**Review by minitu — COMMENTED** [review]
Looks good as an example, though it might be good to merge this in after optimizing the communication routines.

**Comment by ritvikrao** [comment]
Can you add a readme to explain what this examples does and how to run it?

**Comment by ZwFink** [comment]
@ritvikrao

**Comment by ritvikrao** [comment]
The code requires Scipy to be installed, so I added that requirement to the readme.

**Comment by ZwFink** [comment]
Why does it require scipy?

**Comment by ritvikrao** [comment]
When running on my local machine with the example command you provided, I got an error saying I couldn't run without it:

`----------------- Python Stack Traceback PE 0 -----------------
  File "charm4py/charmlib/charmlib_cython.pyx", line 877, in charm4py.charmlib.charmlib_cython.recvArrayMsg
  File "/usr/local/lib/python3.11/site-packages/charm4py/charm.py", line 394, in recvArrayMsg
    em.run(obj, header, args)  # now call the user's array element __init__
    ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/charm4py/entry_method.py", line 61, in _run
    charm.process_em_exc(e, obj, header)
  File "/usr/local/lib/python3.11/site-packages/charm4py/charm.py", line 260, in process_em_exc
    raise e
  File "/usr/local/lib/python3.11/site-packages/charm4py/entry_method.py", line 49, in _run
    ret = getattr(obj, self.name)(*args)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ritvik/charm4py/examples/cannon/cannon.py", line 37, in __init__
    matmul(warmup_c, self.sub_a, self.sub_b)
  File "/usr/local/lib/python3.11/site-packages/numba/core/dispatcher.py", line 442, in _compile_for_args
    raise e
  File "/usr/local/lib/python3.11/site-packages/numba/core/dispatcher.py", line 375, in _compile_for_args
    return_val = self.compile(tuple(argtypes))
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/numba/core/dispatcher.py", line 905, in compile
    cres = self._compiler.compile(args, return_type)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/numba/core/dispatcher.py", line 80, in compile
    status, retval = self._compile_cached(args, return_type)
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/numba/core/dispatcher.py", line 94, in _compile_cached
    retval = self._compile_core(args, return_type)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/numba/core/dispatcher.py", line 107, in _compile_core
    cres = compiler.compile_extra(self.targetdescr.typing_context,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/numba/core/compiler.py", line 744, in compile_extra
    return pipeline.compile_extra(func)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/numba/core/compiler.py", line 438, in compile_extra
    return self._compile_bytecode()
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/numba/core/compiler.py", line 506, in _compile_bytecode
    return self._compile_core()
           ^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/numba/core/compiler.py", line 481, in _compile_core
    raise e
  File "/usr/local/lib/python3.11/site-packages/numba/core/compiler.py", line 472, in _compile_core
    pm.run(self.state)
  File "/usr/local/lib/python3.11/site-packages/numba/core/compiler_machinery.py", line 364, in run
    raise e
  File "/usr/local/lib/python3.11/site-packages/numba/core/compiler_machinery.py", line 356, in run
    self._runPass(idx, pass_inst, state)
  File "/usr/local/lib/python3.11/site-packages/numba/core/compiler_lock.py", line 35, in _acquire_compile_lock
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/numba/core/compiler_machinery.py", line 311, in _runPass
    mutated |= check(pss.run_pass, internal_state)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/numba/core/compiler_machinery.py", line 273, in check
    mangled = func(compiler_state)
              ^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/numba/core/typed_passes.py", line 112, in run_pass
    typemap, return_type, calltypes, errs = type_inference_stage(
                                            ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/numba/core/typed_passes.py", line 93, in type_inference_stage
    errs = infer.propagate(raise_errors=raise_errors)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/numba/core/typeinfer.py", line 1083, in propagate
    errors = self.constraints.propagate(self)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/numba/core/typeinfer.py", line 182, in propagate
    raise e
  File "/usr/local/lib/python3.11/site-packages/numba/core/typeinfer.py", line 160, in propagate
    constraint(typeinfer)
  File "/usr/local/lib/python3.11/site-packages/numba/core/typeinfer.py", line 709, in __call__
    self.resolve(typeinfer, typeinfer.typevars, fnty=fnty)
  File "/usr/local/lib/python3.11/site-packages/numba/core/typeinfer.py", line 606, in resolve
    sig = typeinfer.resolve_call(fnty, pos_args, kw_args)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/numba/core/typeinfer.py", line 1577, in resolve_call
    return self.context.resolve_function_type(fnty, pos_args, kw_args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/numba/core/typing/context.py", line 196, in resolve_function_type
    res = self._resolve_user_function_type(func, args, kws)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/numba/core/typing/context.py", line 248, in _resolve_user_function_type
    return func.get_call_type(self, args, kws)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/numba/core/types/functions.py", line 312, in get_call_type
    raise e
  File "/usr/local/lib/python3.11/site-packages/numba/core/types/functions.py", line 308, in get_call_type
    sig = temp.apply(nolitargs, nolitkws)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/numba/core/typing/templates.py", line 350, in apply
    sig = generic(args, kws)
          ^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/numba/core/typing/templates.py", line 613, in generic
    disp, new_args = self._get_impl(args, kws)
                     ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/numba/core/typing/templates.py", line 712, in _get_impl
    impl, args = self._build_impl(cache_key, args, kws)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/numba/core/typing/templates.py", line 816, in _build_impl
    disp_type.get_call_type(self.context, args, kws)
  File "/usr/local/lib/python3.11/site-packages/numba/core/types/functions.py", line 541, in get_call_type
    self.dispatcher.get_call_template(args, kws)
  File "/usr/local/lib/python3.11/site-packages/numba/core/dispatcher.py", line 318, in get_call_template
    self.compile(tuple(args))
  File "/usr/local/lib/python3.11/site-packages/numba/core/dispatcher.py", line 905, in compile
    cres = self._compiler.compile(args, return_type)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/numba/core/dispatcher.py", line 80, in compile
    status, retval = self._compile_cached(args, return_type)
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/numba/core/dispatcher.py", line 94, in _compile_cached
    retval = self._compile_core(args, return_type)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/numba/core/dispatcher.py", line 107, in _compile_core
    cres = compiler.compile_extra(self.targetdescr.typing_context,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/numba/core/compiler.py", line 744, in compile_extra
    return pipeline.compile_extra(func)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/numba/core/compiler.py", line 438, in compile_extra
    return self._compile_bytecode()
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/numba/core/compiler.py", line 506, in _compile_bytecode
    return self._compile_core()
           ^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/numba/core/compiler.py", line 481, in _compile_core
    raise e
  File "/usr/local/lib/python3.11/site-packages/numba/core/compiler.py", line 472, in _compile_core
    pm.run(self.state)
  File "/usr/local/lib/python3.11/site-packages/numba/core/compiler_machinery.py", line 364, in run
    raise e
  File "/usr/local/lib/python3.11/site-packages/numba/core/compiler_machinery.py", line 356, in run
    self._runPass(idx, pass_inst, state)
  File "/usr/local/lib/python3.11/site-packages/numba/core/compiler_lock.py", line 35, in _acquire_compile_lock
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/numba/core/compiler_machinery.py", line 311, in _runPass
    mutated |= check(pss.run_pass, internal_state)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/numba/core/compiler_machinery.py", line 273, in check
    mangled = func(compiler_state)
              ^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/numba/core/typed_passes.py", line 468, in run_pass
    lower.lower()
  File "/usr/local/lib/python3.11/site-packages/numba/core/lowering.py", line 187, in lower
    self.lower_normal_function(self.fndesc)
  File "/usr/local/lib/python3.11/site-packages/numba/core/lowering.py", line 226, in lower_normal_function
    entry_block_tail = self.lower_function_body()
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/numba/core/lowering.py", line 256, in lower_function_body
    self.lower_block(block)
  File "/usr/local/lib/python3.11/site-packages/numba/core/lowering.py", line 270, in lower_block
    self.lower_inst(inst)
  File "/usr/local/lib/python3.11/site-packages/numba/core/lowering.py", line 448, in lower_inst
    val = self.lower_assign(ty, inst)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/numba/core/lowering.py", line 660, in lower_assign
    return self.lower_expr(ty, value)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/numba/core/lowering.py", line 1196, in lower_expr
    res = self.lower_call(resty, expr)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/numba/core/lowering.py", line 925, in lower_call
    res = self._lower_call_normal(fnty, expr, signature)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/numba/core/lowering.py", line 1167, in _lower_call_normal
    res = impl(self.builder, argvals, self.loc)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/numba/core/base.py", line 1190, in __call__
    res = self._imp(self._context, builder, self._sig, args, loc=loc)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/numba/core/base.py", line 1220, in wrapper
    return fn(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/numba/np/linalg.py", line 548, in _dot2_codegen
    ensure_blas()
  File "/usr/local/lib/python3.11/site-packages/numba/np/linalg.py", line 59, in ensure_blas
    raise ImportError("scipy 0.16+ is required for linear algebra")
------------- Processor 0 Exiting: Called CmiAbort ------------
Reason: ImportError: scipy 0.16+ is required for linear algebra
[0] Stack Traceback:
  [0:0] libcharm.dylib 0x104dbf471 CmiAbort
  [0:1] charmlib_cython.cpython-311-darwin.so 0x101b061eb __pyx_pw_8charm4py_8charmlib_15charmlib_cython_8CharmLib_83CkAbort
  [0:2] Python 0x100c12978 _PyEval_EvalFrameDefault
  [0:3] Python 0x100c17553 _PyEval_Vector
  [0:4] charmlib_cython.cpython-311-darwin.so 0x101afeb20 __pyx_f_8charm4py_8charmlib_15charmlib_cython_recvArrayMsg
  [0:5] libcharm.dylib 0x104c90562 ArrayElemExt::ArrayElemExt(void*)
  [0:6] libcharm.dylib 0x104c840b6 CkDeliverMessageFree
  [0:7] libcharm.dylib 0x104cca7da CkLocRec::invokeEntry(CkMigratable*, void*, int, bool)
  [0:8] libcharm.dylib 0x104c94535 CkArray::insertElement(CkArrayMessage*, CkArrayIndex const&, int*)
  [0:9] libcharm.dylib 0x104c953f7 CkArray::insertInitial(CkArrayIndex const&, void*)
  [0:10] libcharm.dylib 0x104cc7a7f CkArrayMap::populateInitial(int, CkArrayOptions&, void*, CkArray*)
  [0:11] libcharm.dylib 0x104c930a4 CkArray::CkArray(CkArrayOptions&&, CkMarshalledMessage&&)
  [0:12] libcharm.dylib 0x104c9c3cb CkIndex_CkArray::_call_CkArray_marshall1(void*, void*)
  [0:13] libcharm.dylib 0x104c840b6 CkDeliverMessageFree
  [0:14] libcharm.dylib 0x104c84b0f CkCreateLocalGroup
  [0:15] libcharm.dylib 0x104c855cd CkCreateGroup
  [0:16] libcharm.dylib 0x104c9c0a0 CProxy_CkArray::ckNew(CkArrayOptions const&, CkMarshalledMessage const&, CkEntryOptions const*)
  [0:17] libcharm.dylib 0x104c9154d CkCreateArray(CkArrayMessage*, int, CkArrayOptions)
  [0:18] libcharm.dylib 0x104c90fad CProxy_ArrayBase::ckCreateArray(CkArrayMessage*, int, CkArrayOptions const&)
  [0:19] libcharm.dylib 0x104c894d0 CkCreateArrayExt
  [0:20] charmlib_cython.cpython-311-darwin.so 0x101af2d97 __pyx_pw_8charm4py_8charmlib_15charmlib_cython_8CharmLib_35CkCreateArray
  [0:21] Python 0x100c12978 _PyEval_EvalFrameDefault
  [0:22] Python 0x100c17553 _PyEval_Vector
  [0:23] Python 0x100b3f0e8 _PyObject_FastCallDictTstate
  [0:24] Python 0x100b9d0a8 slot_tp_new
  [0:25] Python 0x100b96daa type_call
  [0:26] Python 0x100b3ee28 _PyObject_MakeTpCall
  [0:27] Python 0x100c12a7d _PyEval_EvalFrameDefault
  [0:28] Python 0x100c17553 _PyEval_Vector
  [0:29] Python 0x100b42257 method_vectorcall
  [0:30] Python 0x100c14d2b _PyEval_EvalFrameDefault
  [0:31] Python 0x100c17553 _PyEval_Vector
  [0:32] Python 0x100b422cc method_vectorcall
  [0:33] Python 0x100b3f7a2 _PyVectorcall_Call
  [0:34] _greenlet.cpython-311-darwin.so 0x100283f9f greenlet::UserGreenlet::inner_bootstrap(_greenlet*, _object*)
  [0:35] _greenlet.cpython-311-darwin.so 0x10028377b greenlet::UserGreenlet::g_initialstub(void*)
  [0:36] _greenlet.cpython-311-darwin.so 0x100286567 greenlet::UserGreenlet::g_switch()
  [0:37] _greenlet.cpython-311-darwin.so 0x100287e18 green_switch(_greenlet*, _object*, _object*)
  [0:38] Python 0x100b4b8fa method_vectorcall_VARARGS_KEYWORDS
  [0:39] Python 0x100c12978 _PyEval_EvalFrameDefault
  [0:40] Python 0x100c17553 _PyEval_Vector
  [0:41] charmlib_cython.cpython-311-darwin.so 0x101afad6f __pyx_f_8charm4py_8charmlib_15charmlib_cython_buildMainchare
  [0:42] libcharm.dylib 0x104c88b41 MainchareExt::MainchareExt(CkArgMsg*)
  [0:43] libcharm.dylib 0x104d44b35 _initCharm(int, char**)
  [0:44] libcharm.dylib 0x104dc0ed0 ConverseInit
  [0:45] libcharm.dylib 0x104c887fc StartCharmExt
  [0:46] charmlib_cython.cpython-311-darwin.so 0x101af7514 __pyx_pw_8charm4py_8charmlib_15charmlib_cython_8CharmLib_53start
  [0:47] Python 0x100c12978 _PyEval_EvalFrameDefault
  [0:48] Python 0x100c06d02 PyEval_EvalCode
  [0:49] Python 0x100c5cdaa run_eval_code_obj
  [0:50] Python 0x100c5cd3a run_mod
  [0:51] Python 0x100c5cbc7 pyrun_file
  [0:52] Python 0x100c5c6d2 _PyRun_SimpleFileObject
  [0:53] Python 0x100c5c09b _PyRun_AnyFileObject
  [0:54] Python 0x100c77348 pymain_run_file_obj
  [0:55] Python 0x100c76d3e pymain_run_file
  [0:56] Python 0x100c76730 Py_RunMain
  [0:57] Python 0x100c77624 Py_BytesMain
  [0:58] dyld 0x7ff80e671345 start
Fatal error on PE 0> ImportError: scipy 0.16+ is required for linear algebra`

**Review by lvkale — COMMENTED** [review]
Lacks verification code (to check if it calculated the correct result). I wouldn't hold back merging it, but consider it. 
Secondly, the run command in README is meant only for user's workstation (laptop etc.), right? May need a comment saying that and to refer to the documentation on how to run on clusters.

**Review by lvkale — COMMENTED** [review]
Another minor (but better to fix in an exemplar program) inconsistency: 
mult_done_future has future in its name but init_done does not.

**Review by ritvikrao — APPROVED** [review]
Looks good to me. Verification code would be helpful, but that can be another PR.

**Pull request merged**